src/core/public/keyboard_shortcut/keyboard_shortcut_service.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Any modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+import { KeyboardShortcutSetup, KeyboardShortcutStart, ShortcutDefinition } from './types';
+
+/**
+ * @internal
+ * @experimental
+ */
+export class KeyboardShortcutService {
+  private shortcuts = new Map<string, ShortcutDefinition[]>();
+
+  public setup(): KeyboardShortcutSetup {
+    return {
+      register: (shortcuts) => this.register(shortcuts),
+    };
+  }
+
+  public start(): KeyboardShortcutStart {
+    this.startEventListener();
+
+    return {
+      register: (shortcuts) => this.register(shortcuts),
+      unregister: (id) => this.unregister(id),
+    };
+  }
+
+  public stop() {
+    this.stopEventListener();
+    this.shortcuts.clear();
+  }
+
+  private normalizeKeyboardShortcutString = (str: string): string => str.toLowerCase();
+
+  private getNamespacedIdForKeyboardShortcut = (
+    shortcut: Pick<ShortcutDefinition, 'id' | 'pluginId'>
+  ) =>
+    `${this.normalizeKeyboardShortcutString(shortcut.id)}.${this.normalizeKeyboardShortcutString(
+      shortcut.pluginId
+    )}`;
+
+  private getEventKeyString = (event: KeyboardEvent): string => {
+    return `${event.ctrlKey ? 'ctrl+' : ''}${event.altKey ? 'alt+' : ''}${
+      event.shiftKey ? 'shift+' : ''
+    }${event.metaKey ? 'cmd+' : ''}${this.normalizeKeyboardShortcutString(event.key)}`;
```

Comment:
for readability can we build a string instead?

let key = '';
if (event.ctrlKey) {
    key += 'shift+';
 }


etc